crates/runtime/src/table.rs

```diff
@@ -43,27 +111,58 @@ impl Table {
                 TableElements::ExternRefs(vec![None; min])
             }
         });
+
+        match plan.style {
+            TableStyle::CallerChecksSignature => Self {
+                storage: TableStorage::Dynamic {
+                    elements,
+                    maximum: plan.table.maximum,
+                },
+            },
+        }
+    }
+
+    /// Create a new static (immovable) table instance for the specified table plan.
+    pub fn new_static(plan: &TablePlan, data: *mut u8, maximum: u32) -> Self {
         match plan.style {
             TableStyle::CallerChecksSignature => Self {
-                elements,
-                maximum: plan.table.maximum,
+                storage: TableStorage::Static {
+                    data,
+                    size: Cell::new(plan.table.minimum),
+                    ty: plan.table.ty.clone(),
+                    maximum: min(plan.table.maximum.unwrap_or(maximum), maximum),
+                },
             },
         }
     }
 
     /// Returns the type of the elements in this table.
     pub fn element_type(&self) -> TableElementType {
-        match &*self.elements.borrow() {
-            TableElements::FuncRefs(_) => TableElementType::Func,
-            TableElements::ExternRefs(_) => TableElementType::Val(crate::ref_type()),
+        match &self.storage {
+            TableStorage::Static { ty, .. } => *ty,
+            TableStorage::Dynamic { elements, .. } => match &*elements.borrow() {
+                TableElements::FuncRefs(_) => TableElementType::Func,
+                TableElements::ExternRefs(_) => TableElementType::Val(crate::ref_type()),
+            },
         }
     }
 
     /// Returns the number of allocated elements.
     pub fn size(&self) -> u32 {
-        match &*self.elements.borrow() {
-            TableElements::FuncRefs(x) => x.len().try_into().unwrap(),
-            TableElements::ExternRefs(x) => x.len().try_into().unwrap(),
+        match &self.storage {
+            TableStorage::Static { size, .. } => size.get(),
+            TableStorage::Dynamic { elements, .. } => match &*elements.borrow() {
+                TableElements::FuncRefs(x) => x.len().try_into().unwrap(),
+                TableElements::ExternRefs(x) => x.len().try_into().unwrap(),
+            },
+        }
+    }
+
+    /// Returns the maximum number of elements.
+    pub fn maximum(&self) -> Option<u32> {
+        match &self.storage {
+            TableStorage::Static { maximum, .. } => Some(*maximum),
```

Comment:
Oh that's a good point about `Table::new`, but I also feel like this is a bit interesting wrt defined tables in wasm modules. This means that the pooling allocator will change the reported limits on memories/tables from instantiated modules relative to the on-demand allocator. That to me feels a bit odd because I figured the limits would reflect the in-wasm limits rather than the system-imposed limits.

Reply:
I believe the instance's type (i.e. the fact that an exported table/memory's maximum would be unbounded vs. artificially constrained by the pooling allocator) is based on the table/memory plan from the `Module` and not from the runtime table/memory.  So Wasmtime API users would see the same limits as reported by the module for any exported tables or memories when using the on-demand allocator vs. the pooling allocator.

This function, `Table::maximum` is only be called when attempting to grow the table.  Thus this is accurately reporting how far the runtime representation of the table can grow before it fails, which to me makes sense for the runtime's representation to report as constrained by the pooling allocator.

Reply:
Aha ok then this already has the semantics I was hoping for, which is that the `Limits` as reported by wasmtime are always those written in the wasm module.